### DISCUSSION TITLE
How to find specified golang function argument field offset precisely?

### DISCUSSION THREAD

**Discussion by g0194776** [bpftrace/bpftrace#1865] [Q&A] [answered]
Hi Guys,

I want to use bpftrace to trace dockerd process. 

For example, I would like to inspect field `ID` value when I executing `docker ps` command. So I decide to set an uprobe to function `reducePsContainer` in dockerd. But, the first argument of this method is a very complex structure, How can I quickly find the specified field offset here?

```go
//Method Declaration
func (daemon *Daemon) reducePsContainer(container *container.Snapshot, ctx *listContext, reducer containerReducer) (*types.Container, error) 

// Snapshot is a read only view for Containers. It holds all information necessary to serve container queries in a
// versioned ACID in-memory store.
type Snapshot struct {
	types.Container

	// additional info queries need to filter on
	// preserve nanosec resolution for queries
	CreatedAt    time.Time
	StartedAt    time.Time
	Name         string
	Pid          int
	ExitCode     int
	Running      bool
	Paused       bool
	Managed      bool
	ExposedPorts nat.PortSet
	PortBindings nat.PortSet
	Health       string
	HostConfig   struct {
		Isolation string
	}
}

type Container struct {
	ID         string `json:"Id"` //I want to inspect this value
	Names      []string
	Image      string
	ImageID    string
	Command    string
	Created    int64
	Ports      []Port
	SizeRw     int64 `json:",omitempty"`
	SizeRootFs int64 `json:",omitempty"`
	Labels     map[string]string
	State      string
	Status     string
	HostConfig struct {
		NetworkMode string `json:",omitempty"`
	}
	NetworkSettings *SummaryNetworkSettings
	Mounts          []MountPoint
}
```

**Comment by fbs** [comment, marked as answer]
Currently this is hard to do. What we need is some kind of DWARF support (or whatever go uses) so we can extract field offsets

**Reply by g0194776** [reply]
Yep, Although I've already done of it with simple data structure, BUT it's too difficult to calculate it precisely, especially for Golang MAP, It's horrible thing when I have to inspect the value I want.

**Reply by g0194776** [reply]
@fbs BTW, I would like to know that how much more time does bpftrace needs to support DWARF?

**Reply by fbs** [reply]
Not any time soon I'm afraid. We're a small team and features are usually added because someone contributes it. DWARF is on my list, but it's far far down.

We happily accept contributions :). It doesn't have to be code, if you do some work to figure out how the feature should work (e.g. how do you use it ?) and write up a proposal that would be really helpful too

**Reply by g0194776** [reply]
@fbs Yep, you are right, I'll think about it. 👍
